internal/ethapi/api.go

```diff
@@ -997,13 +997,85 @@ func (s *PublicBlockChainAPI) Call(ctx context.Context, args CallArgs, blockNrOr
 }
 
 func DoEstimateGas(ctx context.Context, b Backend, args CallArgs, blockNrOrHash rpc.BlockNumberOrHash, gasCap *big.Int) (hexutil.Uint64, error) {
-	// Retrieve the block to act as the gas ceiling
-	block, err := b.BlockByNumberOrHash(ctx, blockNrOrHash)
-	if err != nil {
-		return 0, err
+	// Binary search the gas requirement, as it may be higher than the amount used
+	var (
+		lo  uint64 = params.TxGas - 1
+		hi  uint64
+		cap uint64
+	)
+	if args.Gas != nil && uint64(*args.Gas) >= params.TxGas {
+		hi = uint64(*args.Gas)
+	} else {
+		// Retrieve the block to act as the gas ceiling
+		block, err := b.BlockByNumberOrHash(ctx, blockNrOrHash)
+		if err != nil {
+			return 0, err
+		}
+		hi = block.GasLimit()
+	}
+	if gasCap != nil && hi > gasCap.Uint64() {
+		log.Warn("Caller gas above allowance, capping", "requested", hi, "cap", gasCap)
+		hi = gasCap.Uint64()
+	}
+	cap = hi
+
+	// Set sender address or use a default if none specified
+	if args.From == nil {
+		if wallets := b.AccountManager().Wallets(); len(wallets) > 0 {
+			if accounts := wallets[0].Accounts(); len(accounts) > 0 {
+				args.From = &accounts[0].Address
+			}
+		}
+	}
+	// Use zero-address if none other is available
+	if args.From == nil {
+		args.From = &common.Address{}
+	}
+	// Create a helper to check if a gas allowance results in an executable transaction
+	executable := func(gas uint64) bool {
+		args.Gas = (*hexutil.Uint64)(&gas)
+
+		_, _, failed, err := DoCall(ctx, b, args, blockNrOrHash, nil, vm.Config{}, 0, gasCap)
+		if err != nil || failed {
+			return false
+		}
+		return true
+	}
+	// Execute the binary search and hone in on an executable gas limit
+	for lo+1 < hi {
+		mid := (hi + lo) / 2
+		if !executable(mid) {
+			lo = mid
+		} else {
+			hi = mid
+		}
+	}
+
+	// Fudging to account for gas required to verify signatures + pass around data.
+	// Specifically, this line accounts for the fact that there's a bit of computation performed in
+	// a "real" transaction that won't be covered by an eth_call:
+	//  1. Going into the OVM_SequencerEntrypoint.
+	//  2. Going into the OVM_ProxyEOA + OVM_ECDSAContractAccount.
+	//  3. Verify signatures in various places.
+	// eth_call skips all of this and therefore won't correctly estimate gas by default. We need to
+	// tweak the gas estimate to account for this discrepancy. Cost is quite high here because of
+	// the EVM limitation that CALL can only pass 63/64 of total gas available -- so most of this
+	// gas isn't actually spent during execution but needs to be provided to avoid a revert.
```

Comment:
How good do we feel about this fudge factor working across the board? It could be turned into a config option but unnecessary if we feel good about it working generally